package/MDAnalysis/lib/_cutil.pyx

```diff
@@ -310,4 +313,71 @@ cdef float _norm(float * a):
     result = 0.0
     for n in range(3):
         result += a[n]*a[n]
-    return sqrt(result)
\ No newline at end of file
+    return sqrt(result)
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def find_fragments(np.int64_t[:] atoms, int[:, :] bondlist):
+    """Calculate distinct fragments from nodes and edges
+
+    Parameters
+    ----------
+    atoms : numpy.ndarray
+       array of nodes
+    bonds : numpy.ndarray
+       array of edges.  Any edges which refer to nodes not in *atoms*
+       will be ignored
+
+    Returns
+    -------
+    fragments : list
+       list of arrays, each containing the indices of a fragment
+
+    .. versionaddded:: 0.19.0
+    """
+    cdef intmap bondmap
+    cdef intset todo, frag_todo, frag_done, frag_next
+    cdef vector[int] this_frag
+    cdef int i, a, b
+
+    # grab record of which atoms I have to process
+    # ie set of all nodes
+    for i in range(atoms.shape[0]):
+        todo.insert(atoms[i])
+    # Process edges into map
+    for i in range(bondlist.shape[0]):
+        a = bondlist[i, 0]
+        b = bondlist[i, 1]
+        # only include edges if both are known nodes
+        if todo.count(a) and todo.count(b):
+            bondmap[a].insert(b)
+            bondmap[b].insert(a)
+
+    frags = []
+
+    while not todo.empty():  # While not all nodes have been done
+        # Start a new fragment
+        frag_todo.clear()
+        frag_done.clear()
+        this_frag.clear()
+        # Grab a start point for next fragment
+        frag_todo.insert(deref(todo.begin()))
+        # Loop until fragment fully exploredt
+        while not frag_todo.empty():
+            for a in frag_todo:
+                frag_done.insert(a)
+                this_frag.push_back(a)
+                for b in bondmap[a]:
+                    if not frag_done.count(b):
+                        frag_next.insert(b)
+            # update todo list
+            frag_todo.clear()
```

Comment:
this is kind of ugly, but I wanted to not change `frag_todo` whilst I was iterating over it, but maybe it's safe to do so?  Ideas welcome

Reply:
Since the swap is done in constant time, I am fine with this approach.